Pulumi submission

xer0x opened the pull request:
# Description

This submission uses Pulumi to deploy an NGINX webserver running on an ECS Fargate cluster. The http service has autoscaling based on CPU utilization. The service is load balanced by an Application Load balancer.

The service supports both HTTP, and HTTPS. The HTTPS listener runs with an AWS ACM certificate that's been verified via DNS. I have a copy running at https://example1.dev.sluglab.com/

# Files and folders

The Pulumi files are:

- [index.ts](https://github.com/xer0x/devops-wci-sample/tree/main/deploy/infrastructure/index.ts) -- creates the ECS Cluster
- [vpc.ts](https://github.com/xer0x/devops-wci-sample/tree/main/deploy/components/vpc.ts) -- basic VPC network, subnets, and routing
- [load_balancer.ts](https://github.com/xer0x/devops-wci-sample/tree/main/deploy/components/load_balancer.ts) -- application load balancer
- [app_container.ts](https://github.com/xer0x/devops-wci-sample/tree/main/deploy/components/image.ts) -- builds and publishes the docker image

# Web server response

```shell
$ curl 'https://example1.dev.sluglab.com/'
<html>
  <head><meta charset="UTF-8">
    <title>Hello Pulumi world!</title>
  </head>
  <body>
      <p>Hello Pulumi world!</p>
  </body>
</html>
```

xer0x committed:
Add notes to README.md

xer0x committed:
Create a cluster -- Humble beginnings

xer0x committed:
Basic http-service

xer0x committed:
We have a cluster, and a service

xer0x committed:
Almost a load balancer

xer0x committed:
Still no load balancer

xer0x committed:
Fix ALB -> ECS -- deployment related?

xer0x committed:
Adjust values for deployment health check

lech-r commented on index.ts:
```diff
@@ -0,0 +1,191 @@
+  dependsOn: [app.appImage]
+});
+
+// Security group for HTTP service (This is probably way to strict)
```
Why do you think it is too strict?

lech-r commented on README.md:
````diff
@@ -17,12 +18,35 @@ Application should display "Hello Pulumi world!" in browser when pointed to ```e
+
+The "@pulumi/docker-build" module builds a container named "latest", at least when following their docs. I got stuck building the container for awhile. The example code includes "caching", but it was not worth figuring out why it did not work.
+
+The async promises code is semi-hidden in Pulumi. This totally confused me. `JSON.stringify({value: pulumi_value.apply(x => x + 1)})`` This code won't run the apply, until after the original value is used.  To fix it, I had to add a`pulumi.all()` call to force it to wait.
````
No worries.  Dealing with [Pulumi inputs and outputs](https://www.pulumi.com/docs/iac/concepts/inputs-outputs/) is an acquired art I'm still learning.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,191 @@
+  description: "Security group for the ECS service",
+  ingress: [
+    {
+      /// swapped this to allow everything
```
This rule needs to allow traffic on port 80 only.  Form the ALB Security Group only.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,191 @@
+      toPort: 0,
+      // securityGroups: [alb.albSecurityGroupId], // Only allow traffic from the ALB
+      //securityGroups: [alb.albSecurityGroupId], // Only allow traffic from the ALB
+      cidrBlocks: ["0.0.0.0/0"],
```
Use ```sourceSecurityGroupId``` rather than ```cidrBlocks``` here.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,191 @@
+    ]
+  }),
+  tags: {
+    "tag-key": "tag-value",
```
Tags are optional.  If you don't have an idea for a useful tag, you don't have to tag. :)

On the other hand, you could define common set of tags constant at the top[ of the file, for example include "project" and "stack" as tags, and then use at least that for every taggable resource.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,191 @@
+/// Attach AWS managed policy for execution tasks (we may need more permissions to fetch configuration values)
+const attach = new aws.iam.RolePolicyAttachment("ecsRolePolicyAttachment", {
+  role: ecsTaskExecutionRole.name,
+  policyArn: "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
```
This is good.  And this is better:
`policyArn: aws.iam.ManagedPolicy.AmazonECSTaskExecutionRolePolicy,`

xer0x committed:
Load balancer with https and custom domain

xer0x committed:
Tidy up 🧹

lech-r commented on vpc.ts:
```diff
@@ -0,0 +1,140 @@
+
+/// Create a VPC
+export const vpc = new aws.ec2.Vpc("wci-vpc", {
+  cidrBlock: "10.0.0.0/16",
```
Hard coded CIDR is not scalable.  How would you handle a situation where you would want to add a second stack (a test stack for example)?

Yes, multiple VPCs with same base CIDR are allowed, even in the same AWS account.  But it is likely to cause various problems down the road.  VPC peering is one example.  It is better to have different and non-overlapping CIDR in each environment (each stack).

lech-r commented on vpc.ts:
```diff
@@ -0,0 +1,140 @@
+});
+
+/// Create a public subnet
+const publicSubnet1 = new aws.ec2.Subnet("public-subnet-2a", {
```
The L. 44-82  section begs refactoring.  DRY principle.

lech-r commented on vpc.ts:
```diff
@@ -0,0 +1,140 @@
+});
+
+/// Associate the public subnet with the public route table
+const publicRouteTableAssociation1 = new aws.ec2.RouteTableAssociation("public-route-table-association-1", {
```
The L. 99-117  section begs refactoring.  DRY principle.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+  networkConfiguration: {
+    assignPublicIp: true,
+    securityGroups: [httpServiceSg.id], // WHAT?!  This can cause a Pulumi AWS API error, but it is not clear why. To fix, comment out assignPublicIp and securityGroups
+    subnets: vpc.publicSubnets.map(subnet => subnet.id),
```
Why is your Fargate service deployed to public subnets?

jonsmorrow commented:
@xer0x Thank you for completing this. Lech and I are reviewing and will reach out to schedule a chat.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+/// Create service to run the hello-world server
+const httpService = new aws.ecs.Service("http", {
+  name: "http",
+  cluster: cluster.id,
```
Minor detail - this should be 
`cluster: cluster.arn,`
(arn, not id)

I know, the examples in [Pulumi aws.ecs.Service doc](https://www.pulumi.com/registry/packages/aws/api-docs/ecs/service/) show id, but the "Inputs" section down below clearly says "ARN of an ECS cluster."

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+  taskDefinition: httpTaskDefinition.arn,
+  desiredCount: 3,
+  networkConfiguration: {
+    assignPublicIp: true,
```
Fargate should not be directly exposed to the internet.  That's what you have the ALB for.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+const httpServiceSg = new aws.ec2.SecurityGroup("ecs-service-sg", {
+  vpcId: vpc.vpc.id,
+  description: "Security group for the ECS service",
+  ingress: [
```
Fargate Security Group ingress should allow port 80 traffic from the ALB only.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+  cpu: "256", // 1024 CPU units == 1 vCPU
+  memory: "512",
+  /// 🤯 pulumi.all() makes this wait, so that apply actually happens on time. (race condition 🏎️)
+  containerDefinitions: pulumi.all([app.imageRef]).apply(([imageRef]) => {
```
LGTM.  pulumi.all.apply chain seems to be pretty standard in this situation.

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+
+/// Network security group to ensure traffic gets from ALB -> ECS Service
+//
+// The ALB -> HTTP service were not talking to each other, because I made the security groups too strict.
```
Pretty sure the reason for not talking was incorrect configuration, but not being "too strict".  Please see the other comments below.

(but this is a minor detail)

lech-r commented on index.ts:
```diff
@@ -0,0 +1,165 @@
+        portMappings: [{
+          containerPort: 80
+        }],
+      }
```
No `logConfiguration` ?

xer0x committed:
Add description

xer0x committed:
🧹 house keeping

xer0x committed:
🧪 Test and Lint

xer0x committed:
Try to delete docker images on pulumi destroy

xer0x committed:
😳 tests are not as easy as i expected

xer0x committed:
🙈 reviewing comments

xer0x committed:
Move Pulumi into deploy folder

xer0x committed:
🚛 refactoring

Moving Pulumi files into /deploy/ folder
Aiming to experiment with class constructors, instead of the current inline format

My problem is that testing the index.ts file imports and creates the
entire infrastructure.

We can test smaller pieces by creating smaller components with classes

xer0x committed:
🏷️ common tags for aws resources

xer0x committed:
🚛 move vpc to components/

Partially migrating the Pulumi infrastructure into 2 folders.

/infrastructure has the original style
/components has new classes

xer0x committed:
🐳 docker cacheFrom works, but not cacheTo

xer0x committed:
🚛 continue pulling resources into components/

xer0x committed:
🚛 refactor container building 🐳

xer0x committed:
🚛 copy cluster into components/

xer0x committed:
✍️ readme notes

xer0x committed:
🧩 Split cluster.ts into smaller pieces

xer0x committed:
💊 make linter happier

xer0x committed:
☕️combining infrastructure/ files

xer0x committed:
♾️ refactored infrastructure/ -> components/

Now all the resources are created by components from the
infrastructure/index.ts file. The resources are the same
as before.

lech-r commented on deploy/components/cluster/role.ts:
```diff
@@ -0,0 +1,54 @@
+    /// Create ecsTaskExecutionRole, used to start/deploy tasks, not by the running task
+    //  It is used to pull Amazon ECR images (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_execution_IAM_role.html)
+    this.ecsTaskExecutionRole = new aws.iam.Role("ecsTaskExecutionRole", {
+      name: "ecsTaskExecutionRole",
```
Are you sure you need a static name here?  Why or why not?

lech-r commented on deploy/components/cluster/service.ts:
```diff
@@ -0,0 +1,186 @@
+      networkConfiguration: {
+        assignPublicIp: true,
+        securityGroups: [httpServiceSg.id], // WHAT?!  This can cause a Pulumi AWS API error, but it is not clear why. To fix, comment out assignPublicIp and securityGroups
+        subnets: pulumi.output(args.publicSubnets).apply(subnets => subnets.map(subnet => subnet.id)),
```
Why is the service in public subnets?

lech-r commented on deploy/components/cluster/service.ts:
```diff
@@ -0,0 +1,186 @@
+          protocol: "-1",
+          fromPort: 0,
+          toPort: 0,
+          cidrBlocks: ["0.0.0.0/0"],
```
Ingress should only be on port 80 from the ALB security group.

lech-r commented on deploy/components/cluster/service.ts:
```diff
@@ -0,0 +1,186 @@
+
+    /// Network security group to ensure traffic gets from ALB -> ECS Service
+    //
+    // The ALB -> HTTP service were not talking to each other, because I made the security groups too strict.
```
Not because too strict.  The service and the ALB security groups need to talk to each other on the service traffic port.  See L. 103.

xer0x committed:
🐳 rename docker image

Avoiding colisions by renaming resources. If we deployed multiple
projects/stages into the same AWS account, then this AWS ECR image
would cause an error. All the projects would try to use the same
wci/hello-world repo, and they would hit a deployment error.

xer0x committed:
🐳 part 2: docker rename

Turns out we have to rename the image & pulumi resource at the same time.

xer0x committed:
🔧 little tweaks
